### PR TITLE
Faster, more space-efficient tutorials

### LINKED ISSUE #665: Faster notebook testing
### Summary

We should make our tutorial notebook tests faster.

### Rationale

Our tutorial notebooks currently take 40+ minutes to test. We hope to greatly expand the number of tutorials in TorchGeo in the future, so we should find ways to improve testing time. Some of these improvements may even translate to other tests as well. If we can get our tutorial notebook tests < 10 min, we could consider testing them on all PRs, Python versions, and platforms.

### Implementation

I've had several ideas for how to improve this:

* Cache download files using [cache](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) action
* Run tests in parallel using [pytest-xdist](https://pypi.org/project/pytest-xdist/)
* Use hidden cells or env vars to shorten number of epochs in [nbmake](https://github.com/treebeardtech/nbmake/issues/72)

### Alternatives

If anyone can think of any other options, let me know!

### Additional information

_No response_

### LINKED ISSUE #1074: Trainers tutorial: switch dataset
### Summary

We should switch the dataset for our [PyTorch Lightning Trainers](https://torchgeo.readthedocs.io/en/stable/tutorials/trainers.html) tutorial.

### Rationale

Our current dataset is downloaded from Radiant Earth, requiring additional dependencies and preventing users from running the tutorial without first signing up for an account and modifying the notebook. It would be better to use a dataset that does not require these things so that the tutorial works out of the box.

### Implementation

Most of our other current tutorials use EuroSAT, so we could use that.

### Alternatives

Should we just choose the smallest and fastest to download dataset and use that for all tutorials?

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR attempts to simplify and speed up our notebook tests, and includes the following changes:

- [x] Use [nbmake variable mocking](https://github.com/treebeardtech/nbmake#mock-out-variables-to-simplify-testing-experimental-) instead of checking pytest env vars
- [x] Use tempfile for a single shared download directory
- [x] Download smaller datasets
- [x] Strip output from notebooks to make smaller files
- [x] Rerun tests if they fail, known transient issues

Closes #665 
Closes #1074

### Before

Numbers from [here](https://github.com/microsoft/torchgeo/actions/runs/4186485631/jobs/7255045628).

| Tutorial | Size | Time |
| --- | --- | --- |
| benchmarking | 1.9G (NAIP) + 717M (Chesapeake) | 34m 26s |
| custom dataset | 331M (Sentinel) | 4m 54s |
| getting started | 1.9G (NAIP) + 717M (Chesapeake) | 10s |
| indices | 331M (Sentinel) | 59s* |
| pretrained weights | 4.7G (EuroSAT) | 1m 25s |
| trainers | 6.4G (Cyclone) | 5m 28s* |
| transforms | 4.7G (EuroSAT) | 2m 37s |
| total | 22G† | 50m* |

_*failed, likely longer if passed_
_†duplicate copies of each download_

### After

Numbers from [here](https://github.com/microsoft/torchgeo/actions/runs/4470446316/jobs/7853990133).

| Tutorial | Size | Time |
| --- | --- | --- |
| benchmarking | 1.9G (NAIP) + 717M (Chesapeake) | 2m 1s |
| custom dataset | 331M (Sentinel) | 3m 11s |
| getting started | 1.9G (NAIP) + 717M (Chesapeake) | 10s |
| indices | 19M (EuroSAT100) | 10s |
| pretrained weights | 19M (EuroSAT100) | 15s |
| trainers | 19M (EuroSAT100) | 13s |
| transforms | 19M (EuroSAT100) | 1m 23s |
| total | 3.0G†† | 7m 34s |

_††shared copies across downloads_

### CODE REVIEW

**Comment by calebrob6** [comment]
This is waiting on the 100 image version of EuroSat right?

**Comment by adamjstewart** [comment]
Yep, will start trying to integrate that now.

**Comment by adamjstewart** [comment]
Notebook tests are passing for the first time in almost a year!

**Comment by calebrob6** [comment]
@adamjstewart, is this good to go?

**Comment by adamjstewart** [comment]
Not yet. Things were working but now they're wonky:

* The notebook test wasn't triggered on the last couple commits
* The notebook tests were failing on the couple commits before that
* Still need to see if cache even works

**Comment by adamjstewart** [comment]
I think this PR is mostly complete. I couldn't get caching working, but we can figure that out another day. A few things remaining I'm concerned about:

* nbmake seems to still be finicky, it crashes if I save the cell output in the trainers tutorial
* reviewing git diffs is impossible, wish there was a way to improve this

Depending on whether the notebook was saved locally or on Colab, the indentation changes, meaning every single line is changed (maybe we can find a JSON autoformatter to fix this?). And saving on Colab adds a ton of extraneous metadata that I don't want. We could remove all outputs, which would fix nbmake and reduce the size of the files, but then you wouldn't be able to see plots without running tutorials. We could also use nbsphinx to generate the outputs, but this would be slow and require downloads and need to happen on every commit. We would at least want to make smaller downloads for the last 3 tutorials.

**Comment by adamjstewart** [comment]
Would also be nice if we could run isort/flake8/pyupgrade on our notebooks. Or even store the notebook in a .py file and autoconvert it to .ipynb on the fly like PyTorch does. But I know less about those.

**Review by isaaccorley — APPROVED** [review]

**Comment by adamjstewart** [comment]
Not sure why the tests are crashing again. They were working fine on Colab. May have to try stripping outputs again.

File: .github/workflows/tutorials.yaml [hunk 1]
```diff
@@ -30,12 +30,10 @@ jobs:
...
       if: steps.cache.outputs.cache-hit != 'true'
       run: |
-        pip install .[datasets,docs,tests]
+        pip install .[docs,tests] planetary_computer pystac pytest-rerunfailures
         pip list
     - name: Run notebook checks
-      env:
-        MLHUB_API_KEY: ${{ secrets.MLHUB_API_KEY }}
-      run: pytest --nbmake --nbmake-timeout=3000 docs/tutorials --durations=10
+      run: pytest --nbmake --durations=10 --reruns=10 docs/tutorials
```
Comment (calebrob6) [inline .github/workflows/tutorials.yaml:36]: Can you explain this?

**Review by calebrob6 — APPROVED** [review]

**Pull request merged**